server/text_generation_server/models/custom_modeling/qwen2_vl.py

```diff
@@ -117,37 +120,60 @@ def forward(
             0
         )
         key = apply_rotary_pos_emb_vision(key.unsqueeze(0), rotary_pos_emb).squeeze(0)
-        # TODO: make use of existing RotatoryPositionEmbedding class
 
-        # create the attention mask
-        attention_mask = torch.zeros(
-            [1, hidden_state.shape[0], hidden_state.shape[0]],
-            device=hidden_state.device,
-            dtype=torch.bool,
-        )
-        # TODO: avoid creating the mask in the forward pass, instead define the largest possible mask and slice it
-
-        # apply the cu_seqlens to the attention mask
-        for i in range(1, len(cu_seqlens)):
-            attention_mask[
-                ...,
-                cu_seqlens[i - 1] : cu_seqlens[i],
-                cu_seqlens[i - 1] : cu_seqlens[i],
-            ] = True
-
-        # transpose for the attention mechanism (batch, seqlen, hidden_dim) -> (seqlen, batch, hidden_dim)
-        query = query.transpose(0, 1)
-        key = key.transpose(0, 1)
-        value = value.transpose(0, 1)
-
-        # apply attention
-        attn_output = F.scaled_dot_product_attention(
-            query, key, value, attention_mask, dropout_p=0.0
-        )
-        attn_output = attn_output.transpose(0, 1)
+        # calc maximum sequence length for any batch
+        max_seqlen = torch.max(cu_seqlens[1:] - cu_seqlens[:-1])
```

Comment:
It may be interesting to compute this value once in the `Qwen2VisionModel.forward` instead.

Reply:
great point, updated in the latest commit to only calculate the `max_seqlen` once!